File: src/packages/sdk/src/runtime/adapters/config-to-dsl.ts
/**
 * Config → DSL Adapter - LAYER 2 Compatibility Bridge
 *
 * Migrates existing evalai.config.json and TestSuite configurations
 * to the new defineEval() DSL without breaking user workflows.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import type { TestSuite, TestSuiteCase, TestSuiteConfig } from "../../testing";
import { createResult, defineEval } from "../eval";
import { createEvalRuntime, disposeActiveRuntime, setActiveRuntime } from "../registry";
import type { EvalSpec } from "../types";

/**
 * Configuration file structure (existing evalai.config.json)
 */
interface EvalAIConfig {
  evaluationId?: string;
  gate?: {
    baseline: string;
    report: string;
  };
  [key: string]: unknown;
}

/**
 * Migration result information
 */
interface MigrationResult {
  success: boolean;
  specsGenerated: number;
  errors: string[];
  warnings: string[];
  outputPath: string;
}

/**
 * Convert TestSuite to defineEval() specifications
 */
export function migrateTestSuiteToDSL(testSuite: TestSuite, outputPath: string): MigrationResult {
  const result: MigrationResult = {
    success: true,
    specsGenerated: 0,
    errors: [],
    warnings: [],
    outputPath,
  };

  try {
    // Create isolated runtime for migration
    const runtime = createEvalRuntime();

    // Use the runtime handle to define specs
    const boundDefineEval = (nameOrConfig: any, executor?: any, options?: any) => {
      // The runtime handle manages the active runtime internally
      const { defineEval } = require("../eval");
      return defineEval(nameOrConfig, executor, options);
    };

    // Get test suite data via public methods
    // Note: We need to access the internal data structure for migration
    // This is a limitation of the current TestSuite design
    const suiteData = extractTestSuiteData(testSuite);

    // Generate DSL file content
    const dslContent = generateDSLFromTestSuiteData(suiteData);

    // Write DSL file
    fs.writeFileSync(outputPath, dslContent, "utf-8");

    result.specsGenerated = suiteData.cases.length;
    result.warnings.push(
      `Migrated ${suiteData.cases.length} test cases from TestSuite to defineEval() DSL`,
    );

    // Cleanup runtime
    disposeActiveRuntime();
  } catch (error) {
    result.success = false;
    result.errors.push(
      `Migration failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  return result;
}

/**
 * Extract data from TestSuite instance
 * This is a workaround for the private properties
 */
function extractTestSuiteData(testSuite: TestSuite): {
  name: string;
  cases: TestSuiteCase[];
  config: TestSuiteConfig;
} {
  // Since TestSuite properties are private, we need to reconstruct from usage
  // This is a limitation that should be addressed in a future version

  // For now, we'll create a basic structure and warn the user
  return {
    name: "test-suite", // Can't access private name
    cases: [], // Can't access private config
    config: {
      cases: [],
      timeout: 30000,
      parallel: true,
      stopOnFailure: false,
      retries: 0,
    },
  };
}

/**
 * Convert evalai.config.json to DSL specifications
 */
export function migrateConfigToDSL(configPath: string, outputPath: string): MigrationResult {
  const result: MigrationResult = {
    success: true,
    specsGenerated: 0,
    errors: [],
    warnings: [],
    outputPath,
  };

  try {
    if (!fs.existsSync(configPath)) {
      result.success = false;
      result.errors.push(`Configuration file not found: ${configPath}`);
      return result;
    }

    const configContent = fs.readFileSync(configPath, "utf-8");
    const config: EvalAIConfig = JSON.parse(configContent);

    // Create isolated runtime for migration
    const runtime = createEvalRuntime();

    // Use the runtime handle to define specs
    const boundDefineEval = (nameOrConfig: any, executor?: any, options?: any) => {
      // The runtime handle manages the active runtime internally
      const { defineEval } = require("../eval");
      return defineEval(nameOrConfig, executor, options);
    };

    // Generate basic DSL structure from config
    const dslContent = generateDSLFromConfig(config);

    // Write DSL file
    fs.writeFileSync(outputPath, dslContent, "utf-8");

    result.specsGenerated = 1; // Basic structure generated
    result.warnings.push(
      "Generated basic DSL structure from evalai.config.json. Manual completion required.",
    );

    // Cleanup runtime
    disposeActiveRuntime();
  } catch (error) {
    result.success = false;
    result.errors.push(
      `Config migration failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  return result;
}

/**
 * Generate DSL code from TestSuite data
 */
function generateDSLFromTestSuiteData(suiteData: {
  name: string;
  cases: TestSuiteCase[];
  config: TestSuiteConfig;
}): string {
  const imports = [
    `// Auto-generated from TestSuite: ${suiteData.name}`,
    `// Generated at: ${new Date().toISOString()}`,
    `// This file replaces the old TestSuite configuration`,
    "",
    `import { defineEval, createResult } from '@pauly4010/evalai-sdk';`,
    "",
  ];

  const specs = suiteData.cases.map((testCase, index) => {
    const specName = testCase.id || `${suiteData.name}-case-${index + 1}`;

    // Generate assertion code
    const assertionCode = testCase.assertions
      ? generateAssertionCode(testCase.assertions)
      : "// No assertions defined";

    return [
      `defineEval("${specName}", async (context) => {`,
      `  // Original input: ${testCase.input}`,
      `  const input = context.input;`,
      `  `,
      `  // TODO: Replace with your actual agent/LLM call`,
      `  const output = await simulateAgent(input);`,
      `  `,
      `  // Assertions:`,
      assertionCode,
      `  `,
      `  return createResult({`,
      `    pass: allAssertionsPassed,`,
      `    score: allAssertionsPassed ? 100 : 0,`,
      `    assertions: assertionResults,`,
      `    metadata: {`,
      `      originalInput: ${JSON.stringify(testCase.input)},`,
      `      originalExpected: ${testCase.expected ? JSON.stringify(testCase.expected) : "undefined"},`,
      `    },`,
      `  });`,
      `}, {`,
      `  description: "Migrated from TestSuite case",`,
      `  tags: ["migrated", "testsuite"],`,
      `});`,
      "",
    ].join("\n");
  });

  const helperFunctions = [
    `// Helper function to simulate your agent/LLM`,
    `// Replace this with your actual implementation`,
    `async function simulateAgent(input: string): Promise<string> {`,
    `  // TODO: Implement your actual agent/LLM call here`,
    `  // For now, return a simple echo as placeholder`,
    `  return \`Agent response to: \${input}\`;`,
    `}`,
    "",
  ];

  return [...imports, ...helperFunctions, ...specs].join("\n");
}

/**
 * Generate DSL code from configuration
 */
function generateDSLFromConfig(config: EvalAIConfig): string {
  return [
    `// Auto-generated from evalai.config.json`,
    `// Generated at: ${new Date().toISOString()}`,
    `// This is a basic DSL structure - complete with your actual evaluations`,
    "",
    `import { defineEval, createResult } from '@pauly4010/evalai-sdk';`,
    "",
    `defineEval("basic-evaluation", async (context) => {`,
    `  const input = context.input;`,
    `  `,
    `  // TODO: Replace with your actual agent/LLM call`,
    `  const output = await simulateAgent(input);`,
    `  `,
    `  // TODO: Add your actual evaluation logic`,
    `  const pass = output.length > 0;`,
    `  const score = pass ? 100 : 0;`,
    `  `,
    `  return createResult({`,
    `    pass,`,
    `    score,`,
    `    metadata: {`,
    `      evaluationId: ${config.evaluationId ? JSON.stringify(config.evaluationId) : "undefined"},`,
    `      input,`,
    `      output,`,
    `    },`,
    `  });`,
    `}, {`,
    `  description: "Basic evaluation migrated from evalai.config.json",`,
    `  tags: ["migrated", "config"],`,
    `});`,
    "",
    `// Helper function to simulate your agent/LLM`,
    `// Replace this with your actual implementation`,
    `async function simulateAgent(input: string): Promise<string> {`,
    `  // TODO: Implement your actual agent/LLM call here`,
    `  return \`Agent response to: \${input}\`;`,
    `}`,
    "",
  ].join("\n");
}

/**
 * Generate assertion code from TestSuite assertions
 */
function generateAssertionCode(assertions: ((output: string) => unknown)[]): string {
  // Since we can't analyze the assertion functions at runtime,
  // we generate placeholder code that users need to complete
  return [
    `  // Original assertions: ${assertions.length} defined`,
    `  // TODO: Manually convert these assertions to evaluation logic:`,
    ...assertions.map((_, index) => `  // assertion ${index + 1}: <convert to evaluation logic>`),
    `  `,
    `  const assertionResults = [];`,
    `  let allAssertionsPassed = true;`,
    `  `,
    `  // TODO: Add your actual assertion logic here`,
    `  // Example:`,
    `  // const containsExpected = output.includes(expected);`,
    `  // assertionResults.push({ name: "contains-expected", passed: containsExpected });`,
    `  // if (!containsExpected) allAssertionsPassed = false;`,
  ].join("\n");
}

/**
 * Discover and migrate all TestSuite configurations in a project
 */
export function migrateProjectToDSL(
  projectRoot: string,
  options: {
    outputDir?: string;
    dryRun?: boolean;
  } = {},
): MigrationResult {
  const result: MigrationResult = {
    success: true,
    specsGenerated: 0,
    errors: [],
    warnings: [],
    outputPath: options.outputDir || path.join(projectRoot, ".evalai", "migrated"),
  };

  try {
    // Find evalai.config.json
    const configPath = path.join(projectRoot, "evalai.config.json");
    if (fs.existsSync(configPath)) {
      const outputPath = path.join(result.outputPath, "evalai.config.migrated.ts");

      if (!options.dryRun) {
        fs.mkdirSync(path.dirname(outputPath), { recursive: true });
        const configResult = migrateConfigToDSL(configPath, outputPath);
        result.specsGenerated += configResult.specsGenerated;
        result.errors.push(...configResult.errors);
        result.warnings.push(...configResult.warnings);
      } else {
        result.warnings.push(`Would migrate evalai.config.json to ${outputPath}`);
      }
    }

    // Look for TestSuite usage in TypeScript/JavaScript files
    const testFiles = findTestSuiteFiles(projectRoot);

    for (const testFile of testFiles) {
      const outputPath = path.join(
        result.outputPath,
        path.basename(testFile).replace(/\.(ts|js)$/, ".migrated.ts"),
      );

      if (!options.dryRun) {
        fs.mkdirSync(path.dirname(outputPath), { recursive: true });
        // Note: Actual TestSuite parsing would require AST analysis
        // For now, we create placeholder files
        const placeholderContent = generatePlaceholderDSL(testFile);
        fs.writeFileSync(outputPath, placeholderContent, "utf-8");
        result.specsGenerated += 1;
        result.warnings.push(`Created migration placeholder for ${testFile}`);
      } else {
        result.warnings.push(`Would migrate ${testFile} to ${outputPath}`);
      }
    }

    if (result.specsGenerated === 0) {
      result.warnings.push("No TestSuite configurations found to migrate");
    }
  } catch (error) {
    result.success = false;
    result.errors.push(
      `Project migration failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  return result;
}

/**
 * Find files that might contain TestSuite usage
 */
function findTestSuiteFiles(projectRoot: string): string[] {
  const testFiles: string[] = [];

  function scanDirectory(dir: string) {
    if (!fs.existsSync(dir)) return;

    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);

      if (entry.isDirectory() && !entry.name.startsWith(".") && entry.name !== "node_modules") {
        scanDirectory(fullPath);
      } else if (entry.isFile() && /\.(ts|js)$/.test(entry.name)) {
        try {
          const content = fs.readFileSync(fullPath, "utf-8");
          if (content.includes("createTestSuite") || content.includes("TestSuite")) {
            testFiles.push(fullPath);
          }
        } catch {
          // Skip files that can't be read
        }
      }
    }
  }

  scanDirectory(projectRoot);
  return testFiles;
}

/**
 * Generate placeholder DSL for files that need manual migration
 */
function generatePlaceholderDSL(originalFile: string): string {
  return [
    `// Migration placeholder for: ${originalFile}`,
    `// Generated at: ${new Date().toISOString()}`,
    `// This file contains TestSuite usage that needs manual migration`,
    "",
    `import { defineEval, createResult } from '@pauly4010/evalai-sdk';`,
    "",
    `defineEval("placeholder-from-${path.basename(originalFile)}", async (context) => {`,
    `  // TODO: Manually migrate TestSuite from ${originalFile}`,
    `  const input = context.input;`,
    `  `,
    `  // Replace with your actual evaluation logic`,
    `  const output = await simulateAgent(input);`,
    `  `,
    `  return createResult({`,
    `    pass: output.length > 0,`,
    `    score: output.length > 0 ? 100 : 0,`,
    `    metadata: {`,
    `      migratedFrom: ${JSON.stringify(originalFile)},`,
    `    },`,
    `  });`,
    `}, {`,
    `  description: "Placeholder - complete migration manually",`,
    `  tags: ["placeholder", "needs-migration"],`,
    `});`,
    "",
    `async function simulateAgent(input: string): Promise<string> {`,
    `  // TODO: Implement your actual agent/LLM call`,
    `  return \`Response to: \${input}\`;`,
    `}`,
    "",
  ].join("\n");
}

File: src/packages/sdk-python/src/evalgate_sdk/runtime/adapters/config_to_dsl.py
"""Config → DSL Adapter — LAYER 2 Compatibility Bridge.

Migrates existing evalgate.config.json and TestSuite configurations
to the new define_eval() DSL without breaking user workflows.

Port of ``runtime/adapters/config-to-dsl.ts``.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


@dataclass
class MigrationResult:
    """Migration result information."""

    success: bool = True
    specs_generated: int = 0
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)
    output_path: str = ""


@dataclass
class EvalAIConfig:
    """Configuration file structure (existing evalgate.config.json)."""

    evaluation_id: str | None = None
    gate: dict[str, str] | None = None
    packages: dict[str, Any] | None = None
    extra: dict[str, Any] = field(default_factory=dict)


def migrate_config_to_dsl(
    config_path: str,
    output_path: str,
) -> MigrationResult:
    """Convert evalgate.config.json to DSL specifications."""
    result = MigrationResult(output_path=output_path)

    if not os.path.exists(config_path):
        result.success = False
        result.errors.append(f"Configuration file not found: {config_path}")
        return result

    try:
        with open(config_path, encoding="utf-8") as f:
            config_data = json.load(f)

        config = EvalAIConfig(
            evaluation_id=config_data.get("evaluationId"),
            gate=config_data.get("gate"),
            packages=config_data.get("packages"),
        )

        dsl_content = _generate_dsl_from_config(config)
        Path(output_path).parent.mkdir(parents=True, exist_ok=True)
        Path(output_path).write_text(dsl_content, encoding="utf-8")

        result.specs_generated = 1
        result.warnings.append("Generated basic DSL structure from evalgate.config.json. Manual completion required.")
    except Exception as exc:
        result.success = False
        result.errors.append(f"Config migration failed: {exc}")

    return result


def migrate_testsuite_to_dsl(
    suite_data: dict[str, Any],
    output_path: str,
) -> MigrationResult:
    """Convert TestSuite data to defineEval() DSL specifications."""
    result = MigrationResult(output_path=output_path)

    try:
        name = suite_data.get("name", "test-suite")
        cases = suite_data.get("cases", [])

        dsl_content = _generate_dsl_from_suite(name, cases)
        Path(output_path).parent.mkdir(parents=True, exist_ok=True)
        Path(output_path).write_text(dsl_content, encoding="utf-8")

        result.specs_generated = len(cases)
        result.warnings.append(f"Migrated {len(cases)} test cases from TestSuite to define_eval() DSL")
    except Exception as exc:
        result.success = False
        result.errors.append(f"Migration failed: {exc}")

    return result


def migrate_project_to_dsl(
    project_root: str,
    output_dir: str | None = None,
    dry_run: bool = False,
) -> MigrationResult:
    """Discover and migrate all configurations in a project."""
    out_dir = output_dir or os.path.join(project_root, ".evalgate", "migrated")
    result = MigrationResult(output_path=out_dir)

    try:
        # Find config files
        config_names = [
            "evalgate.config.json",
            "evalai.config.json",
        ]
        config_path: str | None = None
        for name in config_names:
            candidate = os.path.join(project_root, name)
            if os.path.exists(candidate):
                config_path = candidate
                break

        if config_path:
            output_path = os.path.join(out_dir, "evalgate.config.migrated.py")
            if not dry_run:
                sub = migrate_config_to_dsl(config_path, output_path)
                result.specs_generated += sub.specs_generated
                result.errors.extend(sub.errors)
                result.warnings.extend(sub.warnings)
            else:
                result.warnings.append(f"Would migrate {config_path} to {output_path}")

        # Find TestSuite usage in Python files
        test_files = _find_testsuite_files(project_root)
        for test_file in test_files:
            output_path = os.path.join(
                out_dir,
                Path(test_file).stem + ".migrated.py",
            )
            if not dry_run:
                Path(output_path).parent.mkdir(parents=True, exist_ok=True)
                content = _generate_placeholder_dsl(test_file)
                Path(output_path).write_text(content, encoding="utf-8")
                result.specs_generated += 1
                result.warnings.append(f"Created migration placeholder for {test_file}")
            else:
                result.warnings.append(f"Would migrate {test_file} to {output_path}")

        if result.specs_generated == 0:
            result.warnings.append("No TestSuite configurations found to migrate")
    except Exception as exc:
        result.success = False
        result.errors.append(f"Project migration failed: {exc}")

    return result


def _find_testsuite_files(project_root: str) -> list[str]:
    """Find files that might contain TestSuite usage."""
    found: list[str] = []
    for root, dirs, files in os.walk(project_root):
        # Skip hidden dirs and common non-source dirs
        dirs[:] = [
            d for d in dirs if not d.startswith(".") and d not in ("node_modules", "__pycache__", ".git", "venv")
        ]
        for fname in files:
            if not fname.endswith(".py"):
                continue
            full = os.path.join(root, fname)
            try:
                text = Path(full).read_text(encoding="utf-8", errors="ignore")
                if "create_test_suite" in text or "TestSuite" in text:
                    found.append(full)
            except OSError:
                pass
    return found


def _generate_dsl_from_config(config: EvalAIConfig) -> str:
    """Generate DSL code from configuration."""
    now = datetime.now(timezone.utc).isoformat()
    eval_id = json.dumps(config.evaluation_id) if config.evaluation_id else "None"
    return f"""# Auto-generated from evalgate.config.json
# Generated at: {now}
# This is a basic DSL structure — complete with your actual evaluations

from evalgate_sdk import define_eval, create_result


define_eval("basic-evaluation", _basic_eval)


async def _basic_eval(ctx):
    input_text = ctx.input

    # TODO: Replace with your actual agent/LLM call
    output = f"Agent response to: {{input_text}}"

    passed = len(output) > 0
    score = 100 if passed else 0

    return create_result(
        passed=passed,
        score=score,
        output=output,
        metadata={{
            "evaluation_id": {eval_id},
            "input": input_text,
        }},
    )
"""


def _generate_dsl_from_suite(name: str, cases: list[dict[str, Any]]) -> str:
    """Generate DSL code from TestSuite data."""
    now = datetime.now(timezone.utc).isoformat()
    lines = [
        f"# Auto-generated from TestSuite: {name}",
        f"# Generated at: {now}",
        "# This file replaces the old TestSuite configuration",
        "",
        "from evalgate_sdk import define_eval, create_result",
        "",
    ]

    for i, case in enumerate(cases):
        case_id = case.get("id", f"{name}-case-{i + 1}")
        case_input = json.dumps(case.get("input", ""))
        case_expected = json.dumps(case.get("expected"))
        lines.extend(
            [
                f'define_eval("{case_id}", lambda ctx: _eval_{i}(ctx))',
                "",
                f"async def _eval_{i}(ctx):",
                f"    # Original input: {case_input}",
                f"    # Original expected: {case_expected}",
                "    input_text = ctx.input",
                "    # TODO: Replace with your actual agent/LLM call",
                '    output = f"Agent response to: {input_text}"',
                "    passed = len(output) > 0",
                "    return create_result(passed=passed, score=100 if passed else 0, output=output)",
                "",
            ]
        )

    return "\n".join(lines)


def _generate_placeholder_dsl(original_file: str) -> str:
    """Generate placeholder DSL for files that need manual migration."""
    now = datetime.now(timezone.utc).isoformat()
    basename = Path(original_file).stem
    return f"""# Migration placeholder for: {original_file}
# Generated at: {now}
# This file contains TestSuite usage that needs manual migration

from evalgate_sdk import define_eval, create_result


define_eval("placeholder-from-{basename}", _placeholder_eval)


async def _placeholder_eval(ctx):
    # TODO: Manually migrate TestSuite from {original_file}
    input_text = ctx.input
    output = f"Response to: {{input_text}}"
    return create_result(
        passed=len(output) > 0,
        score=100 if len(output) > 0 else 0,
        metadata={{"migrated_from": {json.dumps(original_file)}}},
    )
"""
